### PR TITLE
Crypto scalar multiplication runs in GraalVM

### PR DESCRIPTION
**Pull request by jbsv**

### CODE REVIEW

File: goland/benchmark/Report.md [hunk 1]
```diff
@@ -154,12 +154,13 @@ based on an average over 6 runs.
...
-| WASM C             | 1617968   | 3240597    | 16243341  | 147322609  | 1404344166 | 14450936473 | 177133558246 |
+| [ns/op]            | 1e0       | 1e1       | 1e2       | 1e3         | 1e4         | 1e5          | 1e6           |
+|--------------------|-----------|-----------|-----------|-------------|-------------|--------------|---------------|
+| Native             |     85'237|    796'121|  7'524'008|   84'807'264|  754'702'346| 7'688'306'495| 84'108'037'580|
+| Unikernel (tcp+fs) |101'907'643|103'753'734|112'467'733|  146'313'097|  340'759'106| 2'314'039'681| 26'744'189'506|
+| Graalvm            |  1'381'074|  1'929'079|  5'026'913|   36'368'261|  287'772'421| 2'573'941'698| 27'242'434'886|
+| EVM (local)        |  3'917'656| 39'468'078|367'289'249|4'143'285'328| OutOfMemory |              |               |
+| WASM C             |  1'617'968|  3'240'597| 16'243'341|  147'322'609|1'404'344'166|14'450'936'473|177'133'558'246|
 
 ## Procedures to setup and run the tests
```
Comment (nkcr) [inline goland/benchmark/Report.md:165]: This section should be updated for graalVM

File: goland/tcp/mod.go [hunk 2]
```diff
@@ -83,3 +84,50 @@ func (hs *Service) Execute(snap store.Snapshot, step execution.Step) (execution.
 		Accepted: true,
 	}, nil
 }
+
+// ExecuteGraalvmScalarMultiply execute a scalar multiplication on stored data
+func (hs *Service) ExecuteGraalvmScalarMultiply(snap store.Snapshot, step execution.Step) (execution.Result, error) {
+	res := execution.Result{}
+
+	storedData, err := snap.Get(storeKey[:])
+	if err != nil || len(storedData) == 0 {
```
Comment (nkcr) [inline goland/tcp/mod.go:93]: you're doing two different checks that should be separate. One is an error in getting the value, the other is checking that the value exist. For the later one, `snap.Get` will return `nil` if the value does not exist. So if you want that check you should add:

```go
if storedData == nil {
```

File: javavm/README.md [hunk 3]
```diff
@@ -62,8 +62,24 @@ jenv shell 17.0.1
 jenv version --verbose
 ```
 
-Note: be aware that you'll have to re-build the `Server.java` using:
+Note: be aware that you'll have to re-build tcp server using whenever you change
```
Comment (nkcr) [inline javavm/README.md:65]:
```suggestion
Note: be aware that you'll have to re-build tcp server whenever you change
```

File: javavm/README.md [hunk 4]
```diff
@@ -62,8 +62,24 @@ jenv shell 17.0.1
...
+ from one env (i.e. jdk) to another (i.e. graalvm).
+
+libsodium installation:
+=======================
+On macOS, use the following to install libsodium in the jvm libraries:
```
Comment (nkcr) [inline javavm/README.md:70]: to be consistent with all the other READMEs:

```suggestion
## libsodium installation

On macOS, use the following to install libsodium in the jvm libraries:
```

File: javavm/graalvm_tcp_server/app/src/main/java/graalvm_tcp_server/App.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+public class App {
+    public static void main(String args[]) throws Exception {
+        final int portNumber = 12347;
+        ServerSocket ss = new ServerSocket(portNumber);
+        System.out.println("TCP server started on port " + portNumber);
+
+        Runtime.getRuntime().addShutdownHook(new Thread() {
+            @Override
+            public void run() {
+                System.out.println("Shutdown Hook called");
```
Comment (nkcr) [inline javavm/graalvm_tcp_server/app/src/main/java/graalvm_tcp_server/App.java:20]: But actually no shutdown of the server ? 😅

File: javavm/graalvm_tcp_server/app/src/main/java/smartcontract/SmartScalarMult.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    SmartCrypto sc = new SmartCrypto();
+
+    for (double i = 0; i < ITERATIONS; i++) {
+      sc.scalarmult_ed25519_base_noclamp(output, input);
+    }
+
+    return output;
+  }
+
+  public final int crypto_core_ed25519_SCALARBYTES = 32;
```
Comment (nkcr) [inline javavm/graalvm_tcp_server/app/src/main/java/smartcontract/SmartScalarMult.java:23]: a very small detail, but the good practice (at least 4 years ago 😁 ) was to put global variables and attributes at the top of the class.

**Review by nkcr — COMMENTED** [review]
Nice results 👍

File: javavm/graalvm_tcp_server/app/src/main/java/smartcontract/SmartScalarMult.java (hunk 6 above)
Comment (jbsv) [reply javavm/graalvm_tcp_server/app/src/main/java/smartcontract/SmartScalarMult.java:23]: Some automagic tool pushed that down the file. I didn't mind as we are dealing with a one small page class. But I'll keep the good practice. Thx.

**Review by nkcr — APPROVED** [review]
LGTM 🚀

**Pull request merged**